compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,457 @@
+
+  Vec<Symbol*> symSet;
+  Vec<SymExpr*> symExprs;
+  Map<Symbol*,Vec<SymExpr*>*> defMap;
+  Map<Symbol*,Vec<SymExpr*>*> useMap;
+
+  collectSymbolSetSymExprVec(fn, symSet, symExprs);
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  findCandidatesInFunc(fn, symSet, udcMap);
+
+  freeDefUseMaps(defMap, useMap);
+}
+
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*> & defMap,
+    Map<Symbol*,Vec<SymExpr*>*> & useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo) {
+
+  if(sym && !(toFnSymbol(sym) || toArgSymbol(sym) || toTypeSymbol(sym))) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar = NULL;
+
+    Vec<SymExpr*>* defs = defMap.get(sym);
+    Vec<SymExpr*>* uses = useMap.get(sym);
+
+    if(defs && defs->n == 1 && uses && uses->n == 1) { // check def-use counts
+      SymExpr* se = defs->first();
+      defPar = se->parentExpr;
+
+      //defPar has to be a move without any coercion
+      CallExpr* ce = toCallExpr(defPar);
+      if(ce) {
+        if(ce->isPrimitive(PRIM_MOVE) || ce->isPrimitive(PRIM_ASSIGN)) {
+          Type* lhsType = ce->get(1)->typeInfo();
+          Type* rhsType = ce->get(2)->typeInfo();
+          if(lhsType == rhsType) {
+            // calls to communication functions are generated during codegen. ie
+            // at this time they are still PRIM_MOVEs. Generated communication
+            // calls return their result in a pointer argument, therefore not
+            // suitable for denormalization. See function definition for more
+            // comments
+            if(! primMoveGeneratesCommCall(ce)) {
+              if(! (lhsType->symbol->hasFlag(FLAG_EXTERN))){
+                if(!lhsType->symbol->hasFlag(FLAG_ATOMIC_TYPE)){
+                  //at this point we now that def is fine
+                  def = ce->get(2);
+
+                  //now check if we need to case it when we move it
+                  if(CallExpr* defCe = toCallExpr(def)) {
+                    if(defCe->isPrimitive() &&
+                        isIntegerPromotionPrimitive(defCe->primitive->tag)) {
+                      if(requiresCast(lhsType)) {
+                        *castTo = lhsType;
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+
+      if(def) {
+        *defOut = def;
+        // we have def now find where the value is used
+        SymExpr* se = uses->first();
+        usePar = se->parentExpr;
+        if(CallExpr* ce = toCallExpr(usePar)) {
+          if( !(ce->isPrimitive(PRIM_ADDR_OF) ||
+                ce->isPrimitive(PRIM_ARRAY_GET) ||
+                ce->isPrimitive(PRIM_GET_MEMBER) ||
+                ce->isPrimitive(PRIM_DEREF) ||
+                ce->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
+                (ce->isPrimitive(PRIM_MOVE) && 
+                 ce->get(1)->typeInfo() !=
+                 ce->get(2)->typeInfo()))) {
```

Comment:
I'd write this section like this:

     if (ce->isPrimitive(PRIM_ADDR_OF) ||
         ce->isPrimitive(PRIM_ARRAY_GET) ||
         ... ) {
         // Don't denormalize; type of use unsupported 
      } else {
        // use is OK for denormalization
        use = se;
      }

I generally prefer to use an empty 'if' block with a comment to having complicated logical expressions. But that's just me.